base/binaryplatforms.jl

```diff
@@ -0,0 +1,999 @@
+        return a == b
+    end
+
+    # Otherwise, do the comparison between the the single version cap and the single version:
+    if a_requested
+        return b <= a
+    else
+        return a <= b
+    end
+end
+
+
+
+"""
+    HostPlatform(p::AbstractPlatform)
+
+Convert a `Platform` to act like a "host"; e.g. if it has a version-bound tag such as
+`"libstdcxx_version" => "3.4.26"`, it will treat that value as an upper bound, rather
+than a characteristic.  `Platform` objects that define artifacts generally denote the
+SDK or version that the artifact was built with, but for platforms, these versions are
+generally the maximal version the platform can support.  The way this transformation
+is implemented is to change the appropriate comparison strategies to treat these pieces
+of data as bounds rather than points in any comparison.
+"""
+function HostPlatform(p::AbstractPlatform)
+    if haskey(p, "os_version")
+        set_compare_strategy!(p, "os_version", compare_version_cap)
+    end
+    if haskey(p, "libstdcxx_version")
+        set_compare_strategy!(p, "libstdcxx_version", compare_version_cap)
+    end
+    return p
+end
+
+"""
+    arch(p::AbstractPlatform)
+
+Get the architecture for the given `Platform` object as a `String`.
+
+# Examples
+```jldoctest
+julia> arch(Platform("aarch64", "Linux"))
+"aarch64"
+
+julia> arch(Platform("amd64", "freebsd"))
+"x86_64"
+```
+"""
+arch(p::AbstractPlatform) = get(tags(p), "arch", nothing)
+
+"""
+    os(p::AbstractPlatform)
+
+Get the operating system for the given `Platform` object as a `String`.
+
+# Examples
+```jldoctest
+julia> os(Platform("armv7l", "Linux"))
+"linux"
+
+julia> os(Platform("aarch64", "macos"))
+"macos"
+```
+"""
+os(p::AbstractPlatform) = get(tags(p), "os", nothing)
+
+# As a special helper, it's sometimes useful to know the current OS at compile-time
+function os()
+    if Sys.iswindows()
+        return "windows"
+    elseif Sys.isapple()
+        return "macos"
+    elseif Sys.isbsd()
+        return "freebsd"
+    else
+        return "linux"
+    end
+end
+
+"""
+    libc(p::AbstractPlatform)
+
+Get the libc for the given `Platform` object as a `String`.  Returns `nothing` on
+platforms with no explicit `libc` choices (which is most platforms).
+
+# Examples
+```jldoctest
+julia> libc(Platform("armv7l", "Linux"))
+"glibc"
+
+julia> libc(Platform("aarch64", "linux"; libc="musl"))
+"musl"
+
+julia> libc(Platform("i686", "Windows"))
+```
+"""
+libc(p::AbstractPlatform) = get(tags(p), "libc", nothing)
+
+"""
+    call_abi(p::AbstractPlatform)
+
+Get the call ABI for the given `Platform` object as a `String`.  Returns `nothing` on
+platforms with no explicit `call_abi` choices (which is most platforms).
+
+# Examples
+```jldoctest
+julia> call_abi(Platform("armv7l", "Linux"))
+"eabihf"
+
+julia> call_Abi(Platform("x86_64", "macos"))
+```
+"""
+call_abi(p::AbstractPlatform) = get(tags(p), "call_abi", nothing)
+
+const platform_namess = Dict(
```

Comment:
Typo I think?

Assuming that there is no intention to dynamically add platforms, `platform_name` below could be written as

```
function platform_name(p::AbstractPlatform)
    n = os(p)
    n === nothing && return "Unknown"
    n == "linux"   ? "Linux"   :
    n == "macos"   ? "macOS"   :
    n == "windows" ? "Windows" :
    n == "freebsd" ? "FreeBSD" :
    error("unhandled os $(repr(n))")
end
```

which avoids having to have the dictionary at all, makes it clear that `platform_namess` is not something that is extended and showes that there is a possible unrecoverable error when something unexpected is returned from `os`. The real hardcore version would be to define with an `@enum PlatformType` the different platforms listed here and make `os` return a `Union{Nothing, PlatformType}`  thus making the error impossible by the type system.

Reply:
> Typo I think?

Yes, thank you!

> Assuming that there is no intention to dynamically add platforms

I think having this escape hatch open is prudent.  While I don't expect to add any more future platforms in the near future, who knows, perhaps we'll add OpenBSD support or something and then we'll need to modify this.  It would be a lot easier if our JLL packages can push something into this dictionary via a Pkg hook or something.  :)